examples/connection_pools/src/main.rs

```diff
@@ -0,0 +1,176 @@
+#![feature(plugin, decl_macro)]
+#![plugin(rocket_codegen)]
+
+// TODO: Fix impls to extract multiple connection pools from state
+// As it stands now, only one database type can be fetched from state.
+
+extern crate rocket;
+extern crate rocket_contrib;
+
+use std::ops::Deref;
+use rocket::http::Status;
+use rocket::request::{self, FromRequest};
+use rocket::{Outcome, Request, State};
+use rocket::config;
+use rocket_contrib::databases::Poolable;
+use rocket_contrib::databases::diesel;
+use rocket_contrib::databases::postgres;
+use rocket_contrib::databases::r2d2;
+use rocket_contrib::databases::r2d2_postgres;
+
+// The two annotations would be part of the proc-macro crate to generate the
+// FromRequest implementation for the specified DB type.
+// #[derive(DbConn)]
+// #[database("my_sqlite_database")]
+struct TempStorage(r2d2::PooledConnection<diesel::r2d2::ConnectionManager<diesel::SqliteConnection>>);
+
+// #[derive(DbConn)]
+// #[database("my_primary_database")]
+struct PrimaryDb(r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>);
+
+// #[derive(DbConn)]
+// #[database("my_warehouse_database")]
+struct WarehouseDb(r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>);
+
+fn main() {
+    let rocket = rocket::ignite();
+    let config = rocket.config().clone();
+    rocket
+        .mount("/", routes![test_handler])
+        .manage(TempStorage::pool(config.clone()))
+        .manage(PrimaryDb::pool(config.clone()))
+        .manage(WarehouseDb::pool(config))
```

Comment:
Cloning the config, while not likely to cause performance penalties on boot, isn't an ideal solution. For now, this is the best I could come up with for something rough like this.

Reply:
I think we want these to be attach fairings instead:

  1. They can fail graciously.
  2. An instance of `Rocket`, which contains the config, gets passed in automatically. Right now, the user would have to pass in `config`, which is pretty unergonomic.
  3. We can do more things in the future, like automatically mounting routes that exposes a management interface to a database, without a breaking change.

Reply:
Sure. I can return to the approach of using attach fairings instead of directly managing Rocket state.

I understand points 1 & 2, but don't quite understand your third point.

Reply:
Oh, sorry. I just mean that, since attach fairings can arbitrary modify the `Rocket` instance, they have significantly more flexibility on what they can do. In particular, I was noting that attach fairings can mount arbitrary routes. We might imagine a future where the contrib databases library can mount some number of routes for each database that allows the developer to manage and explore the database, like Django's management interface. Concretely, the user would type:

```rust
rocket::ignite().attach(TempStorage::fairing()).launch()
```

And we might see a few extra routes be mounted:

```sh
Mounting '/my_sqlite_database':
    => GET '/'
    => GET '/<id>'
    ...
```

And if the user were to visit those routes, Rocket would render in some pretty way the contents of the database. Purely for debugging/introspection. Totally a thing for the future, and I'm really just tossing it out there.